File: backend/src/main/java/com/todo/hulkstore/service/ProductService.java
package com.todo.hulkstore.service;

import com.todo.hulkstore.constants.FileType;
import com.todo.hulkstore.constants.ProductStockCondition;
import com.todo.hulkstore.dto.request.ProductRequest;
import com.todo.hulkstore.dto.response.ProductResponse;

import java.io.PrintWriter;
import java.util.List;

public interface ProductService {

    List<ProductResponse> getAllProducts(ProductStockCondition stockCondition);

    void exportToFile(PrintWriter writer, FileType fileType, ProductStockCondition stockCondition);

    ProductResponse getProductById(Long id);

    ProductResponse createProduct(ProductRequest newProduct);

    ProductResponse updateProduct(Long id, ProductRequest updatedProduct);

    void deleteProductById(Long id);
}

File: frontend/src/app/products/services/product.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams, HttpResponse } from '@angular/common/http';

import { Observable } from 'rxjs';

import { Product, ProductStockCondition, ProductRequest } from '../models';
import { FileType } from 'src/app/common/models';

@Injectable()
export class ProductService {

  private baseUrl = 'http://localhost:8080/api/products';

  constructor(private httpService: HttpClient) { }

  getAllProducts(): Observable<Product[]> {
    return this.httpService.get<Product[]>(this.baseUrl);
  }

  getAvailableProducts(): Observable<Product[]> {
    let params = new HttpParams();
    params = params.append('stock', ProductStockCondition.AVAILABLE);
    return this.httpService.get<Product[]>(this.baseUrl, { params });
  }

  getProductById(id: number): Observable<Product> {
    return this.httpService.get<Product>(`${this.baseUrl}/${id}`);
  }

  createProduct(newProductRequest: ProductRequest): Observable<Product> {
    return this.httpService.post<Product>(this.baseUrl, newProductRequest);
  }

  updateProduct(id: number, updatedProductRequest: ProductRequest): Observable<Product> {
    return this.httpService.put<Product>(`${this.baseUrl}/${id}`, updatedProductRequest);
  }

  deleteProductById(id: number): Observable<void> {
    return this.httpService.delete<void>(`${this.baseUrl}/${id}`);
  }

  download(fileType: FileType, stockCondition: ProductStockCondition): Observable<HttpResponse<Blob>> {
    const url = `${this.baseUrl}/export`;

    let params = new HttpParams();
    params = params.append('format', fileType.format);
    params = params.append('stock', stockCondition);

    return this.httpService.get<Blob>(url, {
      params,
      observe: 'response',
      responseType: 'blob' as 'json'
    });
  }
}
